File: src/utils/mathUtils.ts
import { GameOperation, GameLevel, MathQuestion } from "../types";

export function generateQuestion(
  operation: GameOperation,
  level: GameLevel
): MathQuestion {
  let operandA = 0;
  let operandB = 0;
  let correctAnswer = "";

  const randomInt = (min: number, max: number) => {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  };

  switch (operation) {
    case GameOperation.ADD:
      if (level === GameLevel.BEGINNER) {
        // 1-digit + 1-digit
        operandA = randomInt(1, 9);
        operandB = randomInt(1, 9);
      } else if (level === GameLevel.MEDIUM) {
        // 1-digit + 2-digit
        if (Math.random() < 0.5) {
          operandA = randomInt(1, 9);
          operandB = randomInt(10, 99);
        } else {
          operandA = randomInt(10, 99);
          operandB = randomInt(1, 9);
        }
      } else {
        // 2-digit + 2-digit (HARD)
        operandA = randomInt(10, 99);
        operandB = randomInt(10, 99);
      }
      correctAnswer = (operandA + operandB).toString();
      break;

    case GameOperation.SUB:
      if (level === GameLevel.BEGINNER) {
        // 1-digit - 1-digit (result >= 0)
        operandA = randomInt(1, 9);
        operandB = randomInt(0, operandA);
      } else if (level === GameLevel.MEDIUM) {
        // 2-digit - 1-digit (result >= 0)
        operandA = randomInt(10, 99);
        operandB = randomInt(1, 9);
      } else {
        // 2-digit - 2-digit (result >= 0) (HARD)
        operandA = randomInt(10, 99);
        operandB = randomInt(10, operandA);
      }
      correctAnswer = (operandA - operandB).toString();
      break;

    case GameOperation.MUL:
      if (level === GameLevel.BEGINNER) {
        // 1-digit * 1-digit
        operandA = randomInt(1, 9);
        operandB = randomInt(1, 9);
      } else if (level === GameLevel.MEDIUM) {
        // 1-digit * 2-digit
        if (Math.random() < 0.5) {
          operandA = randomInt(1, 9);
          operandB = randomInt(10, 25); // cap 2-digit to make it playable
        } else {
          operandA = randomInt(10, 25);
          operandB = randomInt(1, 9);
        }
      } else {
        // 2-digit * 2-digit (HARD)
        operandA = randomInt(10, 50); // cap to keep it fun and highly mental math soluble
        operandB = randomInt(10, 30);
      }
      correctAnswer = (operandA * operandB).toString();
      break;

    case GameOperation.DIV:
      if (level === GameLevel.BEGINNER) {
        // Whole result, dividend 1-20, divisor 1-5
        // Generate reverse: compute standard random result and divisor first to avoid residues
        const result = randomInt(1, 4); // so max dividend is 20
        operandB = randomInt(1, 5);
        operandA = result * operandB;
        if (operandA > 20) {
          operandA = randomInt(1, 5) * operandB;
        }
        correctAnswer = Math.floor(operandA / operandB).toString();
      } else if (level === GameLevel.MEDIUM) {
        // Whole result, dividend 10-99, divisor 1-9
        const result = randomInt(2, 20);
        operandB = randomInt(1, 9);
        operandA = result * operandB;
        while (operandA < 10 || operandA > 99) {
          const r = randomInt(2, 25);
          const d = randomInt(2, 9);
          operandA = r * d;
          operandB = d;
        }
        correctAnswer = Math.floor(operandA / operandB).toString();
      } else {
        // Whole or 1 decimal, dividend 10-99, divisor 2-10 (HARD)
        // Let's generate random numbers directly, then compute to 1 decimal
        operandA = randomInt(10, 99);
        operandB = randomInt(2, 10);
        const rawResult = operandA / operandB;
        // round to 1 decimal place: e.g. 5.14 -> 5.1, 5.25 -> 5.3
        const val = Math.round(rawResult * 10) / 10;
        correctAnswer = val.toString();
      }
      break;
  }

  return {
    operandA,
    operandB,
    operator: operation,
    correctAnswer
  };
}

export function getOperatorSymbol(op: GameOperation): string {
  switch (op) {
    case GameOperation.ADD:
      return "+";
    case GameOperation.SUB:
      return "−";
    case GameOperation.MUL:
      return "×";
    case GameOperation.DIV:
      return "÷";
  }
}

File: android/app/src/main/java/com/mathgame/tariktambang/utils/MathUtils.kt
package com.mathgame.tariktambang.utils

import com.mathgame.tariktambang.types.GameLevel
import com.mathgame.tariktambang.types.GameOperation
import com.mathgame.tariktambang.types.MathQuestion
import kotlin.math.roundToInt
import kotlin.random.Random

object MathUtils {

    fun generateQuestion(operation: GameOperation, level: GameLevel): MathQuestion {
        var operandA = 0
        var operandB = 0
        var correctAnswer = ""

        fun randomInt(min: Int, max: Int): Int {
            return Random.nextInt(min, max + 1)
        }

        when (operation) {
            GameOperation.ADD -> {
                when (level) {
                    GameLevel.BEGINNER -> {
                        // 1-digit + 1-digit
                        operandA = randomInt(1, 9)
                        operandB = randomInt(1, 9)
                    }
                    GameLevel.MEDIUM -> {
                        // 1-digit + 2-digit or vice versa
                        if (Random.nextFloat() < 0.5f) {
                            operandA = randomInt(1, 9)
                            operandB = randomInt(10, 99)
                        } else {
                            operandA = randomInt(10, 99)
                            operandB = randomInt(1, 9)
                        }
                    }
                    GameLevel.HARD -> {
                        // 2-digit + 2-digit
                        operandA = randomInt(10, 99)
                        operandB = randomInt(10, 99)
                    }
                }
                correctAnswer = (operandA + operandB).toString()
            }

            GameOperation.SUB -> {
                when (level) {
                    GameLevel.BEGINNER -> {
                        // 1-digit - 1-digit (result >= 0)
                        operandA = randomInt(1, 9)
                        operandB = randomInt(0, operandA)
                    }
                    GameLevel.MEDIUM -> {
                        // 2-digit - 1-digit (result >= 0)
                        operandA = randomInt(10, 99)
                        operandB = randomInt(1, 9)
                    }
                    GameLevel.HARD -> {
                        // 2-digit - 2-digit (result >= 0)
                        operandA = randomInt(10, 99)
                        operandB = randomInt(10, operandA)
                    }
                }
                correctAnswer = (operandA - operandB).toString()
            }

            GameOperation.MUL -> {
                when (level) {
                    GameLevel.BEGINNER -> {
                        // 1-digit * 1-digit
                        operandA = randomInt(1, 9)
                        operandB = randomInt(1, 9)
                    }
                    GameLevel.MEDIUM -> {
                        // 1-digit * 2-digit
                        if (Random.nextFloat() < 0.5f) {
                            operandA = randomInt(1, 9)
                            operandB = randomInt(10, 25)
                        } else {
                            operandA = randomInt(10, 25)
                            operandB = randomInt(1, 9)
                        }
                    }
                    GameLevel.HARD -> {
                        // 2-digit * 2-digit (capped to keep mental math playable)
                        operandA = randomInt(10, 50)
                        operandB = randomInt(10, 30)
                    }
                }
                correctAnswer = (operandA * operandB).toString()
            }

            GameOperation.DIV -> {
                when (level) {
                    GameLevel.BEGINNER -> {
                        // Whole result, dividend 1-20, divisor 1-5
                        val result = randomInt(1, 4)
                        operandB = randomInt(1, 5)
                        operandA = result * operandB
                        if (operandA > 20) {
                            operandA = randomInt(1, 5) * operandB
                        }
                        correctAnswer = (operandA / operandB).toString()
                    }
                    GameLevel.MEDIUM -> {
                        // Whole result, dividend 10-99, divisor 2-9
                        var result = randomInt(2, 20)
                        var d = randomInt(1, 9)
                        var calculated = result * d
                        while (calculated < 10 || calculated > 99) {
                            result = randomInt(2, 25)
                            d = randomInt(2, 9)
                            calculated = result * d
                        }
                        operandA = calculated
                        operandB = d
                        correctAnswer = (operandA / operandB).toString()
                    }
                    GameLevel.HARD -> {
                        // Whole or 1 decimal, dividend 10-99, divisor 2-10 (HARD)
                        operandA = randomInt(10, 99)
                        operandB = randomInt(2, 10)
                        val rawResult = operandA.toFloat() / operandB.toFloat()
                        val rounded = (rawResult * 10f).roundToInt() / 10f
                        correctAnswer = if (rounded % 1 == 0f) {
                            rounded.toInt().toString()
                        } else {
                            rounded.toString()
                        }
                    }
                }
            }
        }

        return MathQuestion(operandA, operandB, operation, correctAnswer)
    }

    fun getOperatorSymbol(op: GameOperation): String {
        return when (op) {
            GameOperation.ADD -> "+"
            GameOperation.SUB -> "−"
            GameOperation.MUL -> "×"
            GameOperation.DIV -> "÷"
        }
    }
}
